### PR TITLE
WIP: fix(android): add Referer header for url redirect load

### PR DESCRIPTION
**Pull request by rocwind**
# Summary
For redirected url loading, current behehavior is we always return `true` to prevent it being loaded in `shouldOverrideUrlLoading`, with `TopShouldStartLoadWithRequestEvent` being sent to RN and then call `webview.loadUrl()` if such redirect loading was approved on RN.
However, the `Referer` header would be different between `webview.loadUrl` and return `false` in `shouldOverrideUrlLoading`. It might cause redirect issues ( #1073 ) if `Referer` matters at the server side.  
The code change inside this PR is to save a `refererUrl` when sending `TopShouldStartLoadWithRequestEvent` to RN and use it to construct a request with `Referer` header later.

### What are the steps to reproduce (after prerequisites)?
test loading `https://s.click.taobao.com/fj2tptv` (url from #1073 ) 

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| Android |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [ ] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)

### CODE REVIEW

**Comment by zhongjie-chen** [comment]
+1

**Comment by jamonholmgren** [comment]
Can you fix conflicts @rocwind ? THanks!

**Comment by rocwind** [comment]
> Can you fix conflicts @rocwind ? THanks!

Sure. It's updated. Could you please help to review it? @jamonholmgren 

I noticed that there were some discussions about to deal with custom headers in #460 and thought it might be able to be addressed by expose the target `referer` header in event for client - if client decided to handle the request on its own. However, I am not quite sure about this solution and would be glad to revert it from this PR if it doesn't look good to you.

**Comment by timboektoe** [comment]
I created this issue several days ago:
https://github.com/react-native-community/react-native-webview/issues/1434

Could it be it is related to? https://github.com/react-native-community/react-native-webview/pull/1200

**Comment by rocwind** [comment]
> I created this issue several days ago:
> #1434
> 
> Could it be it is related to? #1200
@timboektoe 
I think it's possible - given it is related to the redirect case on Android. Would you mind to give this PR a try to see if it addresses your problem?

Actually, we are currently using some sort of workaround at RN JS land to live without this native side code fix. It's basiclly the idea comes from https://github.com/react-native-community/react-native-webview/blob/master/docs/Guide.md#setting-custom-headers - intercepts the redirect load on Android and load it with referer header all by props:
```
const { uri } = props;
const refererRef = useRef(uri);
const [source, setSource] = useState({
    uri
});
const onShouldStartLoadWithRequest = useCallback((request) => {
    const { url: newURL } = request;
    setSource({
        uri: newURL,
        headers: {
            Referer: refererRef.current,
        },
    });
    refererRef.current = newURL;
    return false;
}, []);

const onShouldStartLoadWithRequestProps = Platform.OS === 'android' ? { onShouldStartLoadWithRequest } : {}

return (
    <WebView
        source={source} 
        {...onShouldStartLoadWithRequestProps} 
    />);
```

**Comment by timboektoe** [comment]
> > I created this issue several days ago:
> > #1434
> > Could it be it is related to? #1200
> > @timboektoe
> > I think it's possible - given it is related to the redirect case on Android. Would you mind to give this PR a try to see if it addresses your problem?
> 
> Actually, we are currently using some sort of workaround at RN JS land to live without this native side code fix. It's basiclly the idea comes from https://github.com/react-native-community/react-native-webview/blob/master/docs/Guide.md#setting-custom-headers - intercepts the redirect load on Android and load it with referer header all by props:
> 
> ```
> const { uri } = props;
> const refererRef = useRef(uri);
> const [source, setSource] = useState({
>     uri
> });
> const onShouldStartLoadWithRequest = useCallback((request) => {
>     const { url: newURL } = request;
>     setSource({
>         uri: newURL,
>         headers: {
>             Referer: refererRef.current,
>         },
>     });
>     refererRef.current = newURL;
>     return false;
> }, []);
> 
> const onShouldStartLoadWithRequestProps = Platform.OS === 'android' ? { onShouldStartLoadWithRequest } : {}
> 
> return (
>     <WebView
>         source={source} 
>         {...onShouldStartLoadWithRequestProps} 
>     />);
> ```

Hi I will try out this PR to see if it fixes my problem. Unfortunately the site that exposes the problem is down due to a technical problem. Hopefully I can check it later this week.

**Comment by timboektoe** [comment]
Sorry to bother, but I checked out this PR to see if it can solve my problem, but I get stuck while testing within react-native-webview following these instructions: https://github.com/react-native-community/react-native-webview/blob/master/docs/Contributing.md

The error I get is: error ENOENT: no such file or directory, open 'undefined/src/main/AndroidManifest.xml'. (after running command yarn start:android)

**Comment by timboektoe** [comment]
We have now tested this PR in combination with issue https://github.com/react-native-community/react-native-webview/issues/1434 . Unfortunately this PR does not fix the problem. We further investigated on this and we think that the reason is that the proposed solution does not work for situations where there are multiple consecutive redirects, which is the case in issue 1434:

This is what we see when we monitor the network in the desktop browser.

1. request: https://mijn.uwv.nl/iam/digid, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: https://uwvnederland.queue-it.net/
2. request: https://uwvnederland.queue-it.net/, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: https://mijn.uwv.nl/iam/digid
3. request: https://mijn.uwv.nl/iam/digid, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: /iam/login
4. request: https://mijn.uwv.nl/iam/login, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 200,

In other words when a request is redirected multiple times, the refererer keeps the same. The implemented solution in the PR seems to always copy the current request URL in the referer header which is incorrect in case of a 302. We have tested this assumption using the javascript workaround with 'onShouldStartLoadWithRequestProps' and it works correctly in our scenario, thereby handling our redirect issue on android in a not so elegant way.

**Comment by rocwind** [comment]
> We have now tested this PR in combination with issue #1434 . Unfortunately this PR does not fix the problem. We further investigated on this and we think that the reason is that the proposed solution does not work for situations where there are multiple consecutive redirects, which is the case in issue 1434:
> 
> This is what we see when we monitor the network in the desktop browser.
> 
> 1. request: https://mijn.uwv.nl/iam/digid, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: https://uwvnederland.queue-it.net/
> 2. request: https://uwvnederland.queue-it.net/, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: https://mijn.uwv.nl/iam/digid
> 3. request: https://mijn.uwv.nl/iam/digid, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 302, location: /iam/login
> 4. request: https://mijn.uwv.nl/iam/login, refererer: https://www.uwv.nl/particulieren/mijnuwv/index.aspx, response 200,
> 
> In other words when a request is redirected multiple times, the refererer keeps the same. The implemented solution in the PR seems to always copy the current request URL in the referer header which is incorrect in case of a 302. We have tested this assumption using the javascript workaround with 'onShouldStartLoadWithRequestProps' and it works correctly in our scenario, thereby handling our redirect issue on android in a not so elegant way.

@timboektoe thank you for the testing and feedback, I didn't realize the refererer should keep the same during consecutive redirects. I guess I need to mark this PR as WIP and see if we can figure out a proper solution to this case...

**Comment by rocwind** [comment]
It looks like I can't update this PR anymore since it's already closed.
I think I found a solution to the multiple consecutive redirects case (although it only works for Android N[7.0] or above due to the API it depends on) and opened #1525 with it.
